Core/include/Acts/Geometry/detail/GeometryIdGenerators.hpp

```diff
@@ -0,0 +1,234 @@
+  /// @param volume is the volume for which the geometry Ids
+  /// are set
+  void generateIds(DetectorVolume& volume);
+};
+
+/// @brief A portal counter per volume
+class PortalCounter {
+ public:
+  /// @brief Generate geometry Ids for the objects here
+  /// @param volume is the volume for which the geometry Ids
+  /// are set
+  void generateIds(DetectorVolume& volume);
+};
+
+/// @brief A sensitive counter per volume
+class SensitiveCounter {
+ public:
+  /// @brief Generate geometry Ids for the objects here
+  /// @param volume is the volume for which the geometry Ids
+  /// are set
+  void generateIds(DetectorVolume& volume);
+};
+
+/// @brief A passive counter per volume
+class PassiveCounter {
+ public:
+  /// @brief Generate geometry Ids for the objects here
+  /// @param volume is the volume for which the geometry Ids
+  /// are set
+  void generateIds(DetectorVolume& volume);
+};
+
+/// @brief checks for duplicate geometry ids and
+/// throws exception
+class DuplicateIdChecker {
+ public:
+  // The already 'seen' geometry Ids
+  std::unordered_map<GeometryIdentifier, bool> m_geometryIdMap;
+  // The already 'seen' volumes
+  std::unordered_map<const DetectorVolume*, bool> m_processedVolumes;
+  // The already 'seen' portals
+  std::unordered_map<const Portal*, bool> m_processedPortals;
+
+  /// @brief Checks the geometry ids of all objects and fills
+  /// them into a map, throws an exception if a geometry id is
+  /// assigned twice.
+  ///
+  /// @param volume is the volume for which the geometry Ids
+  /// are checked
+  void generateIds(DetectorVolume& volume) noexcept(false);
+};
+
+/// @brief Checks for unset Geometry Ids and throws exception
+/// in case on eis found
+class UnsetIdChecker {
+ public:
+  /// @brief Checks the geometry ids of all objects and fills
+  /// them into a map, throws an exception if a geometry id is
+  /// assigned twice.
+  ///
+  /// @param volume is the volume for which the geometry Ids
+  /// are checked
+  void generateIds(DetectorVolume& volume) noexcept(false);
+};
+
+/// Volume restricted generator - this is a helper class that
+/// allows to restrict Id Generators to a specific volume
+/// by provind a list of detector volume pointers.
+///
+/// @tparam generator_t the generator type restricted tot his volume(s)
+///
+/// This can be used, e.g. for a layer id generation for
+/// volumes representing different layer volumes
+template <typename generator_t>
+class VolumeRestrictedIdGenerator {
+ public:
+  /// Any type of Id generator
+  generator_t generator;
+  /// The volume restriction for this generator
+  std::vector<const DetectorVolume*> restrictedVolumes = {};
+
+  /// @brief  Convenience constructor with a single generator
+  ///
+  /// @param gen the generator in question
+  VolumeRestrictedIdGenerator(
+      const generator_t& gen,
+      const std::vector<const DetectorVolume*>& rvolumes)
+      : generator(gen), restrictedVolumes(rvolumes) {}
```

Comment:
passing by value and `std::move` might be preferable

Reply:
This, I will leave, because that's shat you would get back as subvolumes from another volume.